dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -500,19 +590,62 @@ FastAddPeerRes FastAddPeer(EngineStoreServerWrap * server, uint64_t region_id, u
         }
         else
         {
+            const auto & settings = server->tmt->getContext().getSettingsRef();
+            auto maybe_elapsed = fap_ctx->tasks_trace->queryElapsed(region_id);
+            RUNTIME_CHECK_MSG(
+                maybe_elapsed.has_value(),
+                "Task not found, region_id={} new_peer_id={}",
+                region_id,
+                new_peer_id);
+            auto elapsed = maybe_elapsed.value();
+            if (elapsed >= 1000 * settings.fap_task_timeout_seconds)
+            {
+                /// NOTE: Make sure FastAddPeer is the only place to cancel FAP phase 1.
+                // If the task is running, we have to wait it return on cancel and clean,
+                // otherwise a later regular may race with this clean.
+                auto prev_state = fap_ctx->tasks_trace->queryState(region_id);
+                LOG_INFO(
+                    log,
+                    "Cancel FAP due to timeout region_id={} new_peer_id={} prev_state={}",
+                    region_id,
+                    new_peer_id,
+                    magic_enum::enum_name(prev_state));
+                if (prev_state == FAPAsyncTasks::TaskState::Running)
+                {
+                    GET_METRIC(tiflash_fap_task_state, type_blocking_cancel_stage).Increment();
+                    [[maybe_unused]] auto result = fap_ctx->tasks_trace->blockedCancelRunningTask(region_id);
+                    GET_METRIC(tiflash_fap_task_state, type_blocking_cancel_stage).Decrement();
+                }
+                else
+                {
+                    [[maybe_unused]] auto s = fap_ctx->tasks_trace->blockedCancelRunningTask(region_id);
+                }
```

Comment:
why here do the same `blockedCancelRunningTask` but increase the `type_blocking_cancel_stage` in only one branch?

Reply:
Because only cancel a Running task is blocking. Consider other 3 states:
1. Finished -- Cancel means immediate fetch and drop result
2. InQueue -- Cancel means setting the cancel flag, so the task will be canceled once it is executed
3. NotExisting -- Obviously

Reply:
Suggest keep it simple because
* There could be some bugs in `AsyncTasks` that make it not as expected
* Keep it simple to simplify the code as adding/subtracting atomic value is not performance critical here

Reply:
The idea here is not to avoid adding/subtracting atomic value, but to record events that could really block cancel. However, point 1 is reasonable. If there is bug in AsyncTasks, log all events could help us to diagnose.